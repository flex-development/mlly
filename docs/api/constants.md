---
head:
  - - meta
    - name: description
      content: Constant values exported from @flex-development/mlly
  - - meta
    - name: keywords
      content:
        - ecmascript
        - es modules
        - esm
        - node
  - - meta
    - name: priority
      content: 0.8
---

# Constants

Constant values exported from `mlly`.

These values are used as defaults for several [library functions](index.md).

<script setup lang="ts">
import { useData } from 'vitepress'
import type ThemeConfig from '../.vitepress/theme/config'
import type Documentation from '../.vitepress/theme/documentation'

const { site } = useData<ThemeConfig>()
const { documentation } = site.value.themeConfig

/**
 * Documentation objects.
 *
 * @const {Documentation[]} docs
 */
const docs: Documentation[] = documentation.filter(doc => {
  return /src\/constants\.ts$/.test(doc.file)
})
</script>

<Doc v-for="doc in docs" :doc="doc.doc" :key="doc.file" />
