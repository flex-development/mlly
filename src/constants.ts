/**
 * @file Constants
 * @module mlly/constants
 */

/**
 * CommonJS syntax regex.
 *
 * @const {RegExp} CJS_SYNTAX_REGEX
 */
export const CJS_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:module\.)?exports(?:\.\w+|(?<!\s+=))|require(?=\(.)|require\.\w+|__dirname|__filename)/gm

/**
 * Dynamic import statement regex.
 *
 * @const {RegExp} DYNAMIC_IMPORT_REGEX
 */
export const DYNAMIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:(?:const|let|var)(?:(?<name>\s+\w+\s*)|\s*{\s*(?<names>\w+(?:,\s*\w+)*)\s*}\s*)(?:\s*=\s*?)?.)?(?:await)?\s+import\s*\((?<expression>(?:[^()]+|\((?:[^()]+|\([^()]*\))*\))*)\)/gm

/**
 * ESM syntax regex.
 *
 * @const {RegExp} ESM_SYNTAX_REGEX
 */
export const ESM_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:export|import)[\s\w*,{}]*(?=\sfrom)|export\b\s*(?:[*{]|async function|(?:abstract\s)?class|const|default|enum|function|interface|let|type|var)|await import|import\.meta\.(?:env(?:\.\w+)?|resolve|url))/gm

/**
 * Static import statement regex.
 *
 * @const {RegExp} STATIC_IMPORT_REGEX
 */
export const STATIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)import\s*(?:[\s"']*(?<imports>[\w\t\n\r $*,/{}]+)from\s*)?["']\s*(?<specifier>(?<="\s*)[^"']*[^\s"'](?=\s*")|(?<=["']\s*)[^']*[^\s'](?=\s*'))\s*["']\s*/gm
