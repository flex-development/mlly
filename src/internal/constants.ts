/**
 * @file Internals - Constants
 * @module mlly/internal/constants
 */

/**
 * CommonJS syntax regex.
 *
 * @internal
 *
 * @const {RegExp} CJS_SYNTAX_REGEX
 */
export const CJS_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:module\.)?exports(?:\.\w+|(?<!\s+=))|require(?=\(.)|require\.\w+|__dirname|__filename|await import)/gm

/**
 * Dynamic import statement regex.
 *
 * @internal
 *
 * @const {RegExp} DYNAMIC_IMPORT_REGEX
 */
export const DYNAMIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:(?:const|let|var)(?:(?<name>\s+\w+\s*)|\s*{\s*(?<names>\w+(?:,\s*\w+)*)\s*}\s*)(?:\s*=\s*?)?.)?(?:await)?\s+import\s*\((?<expression>(?:[^()]+|\((?:[^()]+|\([^()]*\))*\))*)\)/gm

/**
 * ESM syntax regex.
 *
 * @internal
 *
 * @const {RegExp} ESM_SYNTAX_REGEX
 */
export const ESM_SYNTAX_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)((?:export|import)[\s\w*,{}]*(?=\sfrom)|export\b\s*(?:[*{]|async function|(?:abstract\s)?class|const|default|enum|function|interface|let|type|var)|await import|import\.meta\.(?:env(?:\.\w+)?|resolve|url))/gm

/**
 * Declaration export statement regex.
 *
 * **Note**: Captures [declaration type][1] and export name only.
 *
 * [1]: ./types/declaration.ts
 *
 * @const {RegExp} EXPORT_DECLARATION_REGEX
 */
export const EXPORT_DECLARATION_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?: +declare)? +(?<declaration>(?:(?:abstract +)?class|(?:default +)?async +function\*?|(?:default +)?function\*?|default +async|default(?!(?: +async)? +function\*?)|(?:const +)?enum|const|interface|let|namespace|type(?! *{)|var)) +(?<name>[\w$]+(?!.*\b +=>))?/gm

/**
 * Named export statement regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_NAMED_REGEX
 */
export const EXPORT_NAMED_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport(?:\s+type)?\s+{(?<exports>[^}]+?)[\s,]*}(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Aggregate export statement (`export * ...`) regex.
 *
 * @internal
 *
 * @const {RegExp} EXPORT_STAR_REGEX
 */
export const EXPORT_STAR_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)\bexport\s*\*(?:\s*as\s+(?<name>[\w$]+)\s+)?\s*(?:\s*from\s*["']\s*(?<specifier>(?<="\s*)[^"]*[^\s"](?=\s*")|(?<='\s*)[^']*[^\s'](?=\s*'))\s*["'][^\n;]*)?/gm

/**
 * Require statement regex.
 *
 * @internal
 *
 * @const {RegExp} REQUIRE_STATEMENT_REGEX
 */
export const REQUIRE_STATEMENT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)(?:\bconst[ {]+(?<imports>[\w\t\n\r $*,/]+)[ =}]+)?(?<type>\brequire(?:\.resolve)?)\(["'] *(?<specifier>(?<="\s*)[^"']*[^\s"'](?=\s*")|(?<=["']\s*)[^']*[^\s'](?=\s*')) *["']\)/gm

/**
 * Static import statement regex.
 *
 * @internal
 *
 * @const {RegExp} STATIC_IMPORT_REGEX
 */
export const STATIC_IMPORT_REGEX: RegExp =
  /(?<!(?:\/\/|\*).*)import\s*(?:[\s"']*(?<imports>[\w\t\n\r $*,/{}]+)from\s*)?["']\s*(?<specifier>(?<="\s*)[^"]*[^\s":](?=\s*")|(?<=["']\s*)[^']*[^\s':](?=\s*'))\s*["']\s*/gm
