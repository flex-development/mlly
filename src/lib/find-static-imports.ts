/**
 * @file findStaticImports
 * @module mlly/lib/findStaticImports
 */

import { STATIC_IMPORT_REGEX } from '#src/constants'
import type { StaticImport } from '#src/interfaces'

/**
 * Finds all static import statements in `code`.
 *
 * @param {string} code - Code to check
 * @return {StaticImport[]} Static import statement objects
 */
const findStaticImports = (code: string): StaticImport[] => {
  return [...code.matchAll(STATIC_IMPORT_REGEX)].map(match => {
    const { 0: statement = '', index: start = 0, groups = {} } = match
    const { imports = '', specifier = '' } = groups

    /**
     * `import * as <import>` check.
     *
     * @const {boolean} star
     */
    const star: boolean = imports.startsWith('* as')

    return {
      code: statement,
      end: start + statement.length,
      imports: star
        ? [imports.trim()]
        : imports
            .replace(/^type\s*|[{}]/gm, '')
            .split(/,\n?/)
            .map(name => name.trim()),
      specifier,
      start,
      type: star
        ? 'star'
        : /{(?<imports>[\w\t\n\r ,]+)}/gm.test(statement)
        ? 'named'
        : 'default'
    }
  })
}

export default findStaticImports
