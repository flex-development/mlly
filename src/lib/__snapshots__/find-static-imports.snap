// Vitest Snapshot v1

exports[`unit:lib/findStaticImports > should find default static import 1`] = `
[
  {
    "code": "import pb from \\"pretty-bytes\\"",
    "end": 29,
    "imports": [
      "pb",
    ],
    "specifier": "pretty-bytes",
    "start": 0,
    "type": "default",
  },
]
`;

exports[`unit:lib/findStaticImports > should find default type import 1`] = `
[
  {
    "code": "import type Nullable from '@flex-development/tutils/types/nullable'",
    "end": 67,
    "imports": [
      "Nullable",
    ],
    "specifier": "@flex-development/tutils/types/nullable",
    "start": 0,
    "type": "default",
  },
]
`;

exports[`unit:lib/findStaticImports > should find named static imports in multi-line statement 1`] = `
[
  {
    "code": "import {
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
} from './lib'",
    "end": 122,
    "imports": [
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "specifier": "./lib",
    "start": 0,
    "type": "named",
  },
]
`;

exports[`unit:lib/findStaticImports > should find named static imports in single-line statement 1`] = `
[
  {
    "code": "import { readPackage, type PackageJson } from \\"read-pkg\\"",
    "end": 56,
    "imports": [
      "readPackage",
      "type PackageJson",
    ],
    "specifier": "read-pkg",
    "start": 0,
    "type": "named",
  },
]
`;

exports[`unit:lib/findStaticImports > should find named type imports in multi-line statement 1`] = `
[
  {
    "code": "import type {
  BuildOptions,
  BuildResult,
  OutputFile,
  Plugin,
  PluginBuild
} from 'esbuild'",
    "end": 99,
    "imports": [
      "BuildOptions",
      "BuildResult",
      "OutputFile",
      "Plugin",
      "PluginBuild",
    ],
    "specifier": "esbuild",
    "start": 0,
    "type": "named",
  },
]
`;

exports[`unit:lib/findStaticImports > should find named type imports in single-line statement 1`] = `
[
  {
    "code": "import type { Join, Nullable, Overwrite } from '@flex-development/tutils'",
    "end": 73,
    "imports": [
      "Join",
      "Nullable",
      "Overwrite",
    ],
    "specifier": "@flex-development/tutils",
    "start": 0,
    "type": "named",
  },
]
`;

exports[`unit:lib/findStaticImports > should find star static import 1`] = `
[
  {
    "code": "import * as color from \\"colorette\\"",
    "end": 34,
    "imports": [
      "* as color",
    ],
    "specifier": "colorette",
    "start": 0,
    "type": "star",
  },
]
`;
